File: src/digital_marketing/schemas/explain.py
"""解释 DTO。"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class ShapFeature(BaseModel):
    name: str
    feature_value: Any | None = None
    shap_value: float | None = None
    mean_abs_shap: float | None = None


class GlobalExplainData(BaseModel):
    run_id: str
    method: str
    n_samples: int | None = None
    top_features: list[dict[str, Any]] = Field(default_factory=list)


class CustomerExplainRequest(BaseModel):
    customer_id: int | None = None
    features: dict[str, Any] | None = None
    run_id: str | None = None
    top_k: int = 10


class CustomerExplainData(BaseModel):
    run_id: str
    model_name: str | None = None
    method: str
    proba: float | None = None
    label: int | None = None
    threshold: float | None = None
    customer_id: int | None = None
    top_features: list[dict[str, Any]] = Field(default_factory=list)

File: frontend/src/api/explain.ts
import { getData, postData } from './http'

export interface ShapFeature {
  name: string
  feature_value?: unknown
  shap_value?: number
  mean_abs_shap?: number
}

export interface GlobalExplainData {
  run_id: string
  method: string
  n_samples?: number
  top_features: ShapFeature[]
}

export interface CustomerExplainData {
  run_id: string
  model_name?: string
  method: string
  proba?: number
  label?: number
  threshold?: number
  customer_id?: number | null
  top_features: ShapFeature[]
}

export function fetchGlobalExplain(runId?: string) {
  return getData<GlobalExplainData>('/explain/global', runId ? { run_id: runId } : undefined)
}

export function explainCustomer(body: {
  customer_id?: number
  features?: Record<string, unknown>
  run_id?: string
  top_k?: number
}) {
  return postData<CustomerExplainData>('/explain/customer', body)
}
